### ISSUE TITLE
Removed devices not cleaned out

### ISSUE THREAD

**Issue by clockbrain** [mozilla-iot/zigbee-adapter#106]
I've got a few zigbee devices that have stopped working so I remove them from the gateway. Trouble is when I scan for new devices they appear instantly ready to add again but when I add them they still don't work. This occurs whether I reset them or not.

I've checked the config json file and SQLite database and these devices are definitely removed so I can only assume some corrupted config for them is held elsewhere, maybe on the xStick, and not cleaned out when device is removed.

This is what it looks like in the log file:
```
2019-03-08 08:29:28.719 INFO   : zigbee: ----- Nodes -----
2019-03-08 08:29:28.721 INFO   : zigbee: Node: 0013a200415c2377 0000 Name: zb-0013a200415c2377-Dongle       rebindRequired: true endpoints: []
2019-03-08 08:29:28.722 INFO   : zigbee:   Neighbor: 0017880104fc47e8 100c DT: Router R: None     PJ: ? D:  15 LQI: 102
2019-03-08 08:29:28.723 INFO   : zigbee:   Neighbor: 00137a0000040a80 1167 DT: Router R: None     PJ: ? D:  15 LQI: 255
2019-03-08 08:29:28.725 INFO   : zigbee:   Neighbor: 00137a0000040a4b e68f DT: Router R: None     PJ: ? D:  15 LQI: 255
2019-03-08 08:29:28.726 INFO   : zigbee:   Neighbor: 00137a0000040a7f f218 DT: Router R: None     PJ: ? D:  15 LQI: 255
2019-03-08 08:29:28.728 INFO   : zigbee: Node: 00137a0000040a80 1167 Name: zb-00137a0000040a80-smartPlug    rebindRequired: false endpoints: [ '1' ]
2019-03-08 08:29:28.729 INFO   : zigbee: Node: 00137a0000040a4b e68f Name: zb-00137a0000040a4b-smartPlug    rebindRequired: false endpoints: [ '1' ]
2019-03-08 08:29:28.730 INFO   : zigbee: Node: 0017880103cb42ce 63e5 Name: zb-0017880103cb42ce-button       rebindRequired: true endpoints: [ '1', '2' ]
2019-03-08 08:29:28.731 INFO   : zigbee: Node: 0017880103cb3862 1222 Name: zb-0017880103cb3862-button       rebindRequired: true endpoints: [ '1', '2' ]
2019-03-08 08:29:28.732 INFO   : zigbee: Node: 00137a0000040a7f f218 Name: zb-00137a0000040a7f-smartPlug    rebindRequired: false endpoints: [ '1' ]
2019-03-08 08:29:28.733 INFO   : zigbee: Node: 00158d0002c98a27 80f9 Name: zb-00158d0002c98a27-onOffSwitch  rebindRequired: true endpoints: [ '1' ]
2019-03-08 08:29:28.734 INFO   : zigbee: Node: 00158d0002c3afdf e145 Name: zb-00158d0002c3afdf-onOffSwitch  rebindRequired: true endpoints: [ '1' ]
2019-03-08 08:29:28.734 INFO   : zigbee: Node: 00158d0002c3afb8 df29 Name: zb-00158d0002c3afb8-onOffSwitch  rebindRequired: true endpoints: [ '1' ]
2019-03-08 08:29:28.735 INFO   : zigbee: Node: 00158d0002c9865a 2b50 Name: zb-00158d0002c9865a-onOffSwitch  rebindRequired: true endpoints: [ '1' ]
2019-03-08 08:29:28.736 INFO   : zigbee: Node: 0017880104fc47e8 100c Name: zb-0017880104fc47e8-dimmableLigh rebindRequired: true endpoints: [ '11', '242' ]
2019-03-08 08:29:28.737 INFO   : zigbee: -----------------
2019-03-08 08:29:28.738 INFO   : zigbee: ----- Update Network Addresses Complete -----
```

Only the coordinator, first 4 routers, plus one of the Xiaomi switches 00158d0002c9865a exist in my setup and work. Any of the others I try to add again don't work when added.

I also see these sort of errors but I assume it is related:
```
2019-03-08 08:29:27.716 INFO   : zigbee: Sent: Explicit Tx 0017880104fc47e8 ZDO 0000 Network Address Req (0x0000)
2019-03-08 08:29:27.718 INFO   : zigbee: Sent:   Addr:0017880104fc47e8
2019-03-08 08:29:32.555 INFO   : zigbee: Rcvd: ZigBee Transmit Status (0x8B) id: 68 Remote16: fffd Retries: 68 Delivery: Address Not Found (0x24) Discovery: Route Discovery (0x02)
2019-03-08 08:29:32.556 INFO   : zigbee: Transmit Status ERROR: Address Not Found (0x24) id: 68
2019-03-08 08:29:32.557 INFO   : zigbee: { type: 139,
2019-03-08 08:29:32.558 INFO   : zigbee:   id: 68,
2019-03-08 08:29:32.559 INFO   : zigbee:   remote16: 'fffd',
2019-03-08 08:29:32.560 INFO   : zigbee:   transmitRetryCount: 68,
2019-03-08 08:29:32.561 INFO   : zigbee:   deliveryStatus: 36,
2019-03-08 08:29:32.562 INFO   : zigbee:   discoveryStatus: 2 }
```

**Comment by dhylands** [comment]
There are 3 places that information about a device is stored.

- In the database. If you don't see the device in the UI then it shouldn't be in the database anymore.
- The zigbee config file: /home/pi/.mozilla-iot/config/zb-XXXXXXXX.json
- In the routing tables of the devices themselves (typically only powered devices) (including the dongle).

The routing tables are not persisted across a power cycle, so unplugging and replugging the dongle should remove them from the routing table.

If the device is still paired (even though its been removed) then it will wind up back in the routing tables and the zigbee adapter will still query the device to find out attributes. Performing a factory reset on the device will definitely unpair it. We do send out a command to unpair the device when you remove it, but typically only AC powered devices see this. The battery powered devices are typically sleeping when they're not sending messages to the dongle, and they never see the unpair message.

I agree that there should be an easy way to remove "paired but removed" devices from the config file. We need to persist the config information, especially for battery powered devices because they don't talk to the host very frequently, and restarting the gateway would otherwise cause all of the information about them to be forgotten.

**Comment by clockbrain** [comment]
Thanks for the response. Here's the scenario I'm dealing with.

Hue dimmer switch previously worked ok with Things Gateway. Started misbehaving so I tried to remove and re-pair.

The steps I just now followed were:

- Remove Hue Dimmer switch from things gateway via big red button in UI.
- Power off Things Gateway Pi
- Unplug xStick dongle from Pi
- 
- Factory reset Hue Dimmer switch
- Pair Hue dimmer switch with Hue Bridge
- Test - dimmer switch works fine with Hue bridge. Turns on/off Hue Bridge connected Hue lights ok.
- Delete Hue Dimmer from Hue bridge
- Turn off Hue bridge
- 
- Factory reset Hue Dimmer switch
- Plus xStick into Pi and power up Things Gateway
- Test other existing Things zigbee devices (mains switches & Xiaomi entry sensor) - all work ok.
- Enable frames debugging for zigbee adaptor
- Observe this entry in log file which shows that Things Gateway already knows about the dimmer switch before I start pairing.
```
INFO   : zigbee: Node: 0017880103cb3862 1222 Name: zb-0017880103cb3862-button       rebindRequired: true endpoints: [ '1', '2' ]
```
- Tell Things Gateway to scan for new devices. It finds zb-0017880103cb3862-button so add it.
- Test - dimmer switch doesn't work. No response on UI and red led shows on dimmer whenever any button is pressed.
 - Observe this entry in log file which suggests pairing didn't work
```
2019-03-09 01:31:23.584 INFO   : Successfully created new thing zb-0017880103cb3862-button
2019-03-09 01:31:31.977 ERROR  : Thing was not added
2019-03-09 01:31:31.979 ERROR  : addNewThing cancelled
```

I've done this sequence a few times now. No problems with dimmer switch when paired to Hue bridge but doesn't seem to pair properly with the Things Gateway/xStick however many times I try. It appears that Things Gateway has some existing knowledge of that particular dimmer switch which is preventing correct pairing. Power cycling the gateway, unplugging the xStick, and factory resetting the dimmer switch aren't enough to clear it.

Any suggestions on where I go to from here short of reflashing the firmware on the xStick or starting from a fresh Things Gateway image? I'd prefer to not have to do either of those things.

**Comment by dhylands** [comment]
The information about the hue dimmer switch is also stored in the /home/pi/.mozilla-iot/config/zb-XXXXXXXX.json file (the X's will be replaced by the hex digits of the serial number of your zigbee dongle).

To pair the hue remote back with the gateway, factory reset the hue remote, and then press the plus, but don't click save on the item that shows up in the list. This is showing up because it's stored in the config file not because the device is actually paired. If you have the debug frames enabled, you should see some messages from the remote once it's paired. Once you see those messages then its ok to click save. Clicking save will take the dongle back out of pairing mode, and the remote won't get a chance to pair properly.

**Comment by clockbrain** [comment]
That worked! Thanks.

May I respectfully suggest that if there is some way to delay that save button being enabled until pairing is actually complete that would be preferable.

**Comment by clockbrain** [comment]
Just noticed something else. Although my Phillips Hue is paired and works for on/off/dim, it only generates events for the on and off buttons.

I guess this is because mine is a model RWL021 but the classifier code here [https://github.com/mozilla-iot/zigbee-adapter/blob/2eaee692b067ee2d8d56531d7a6dbe9b2659d822/zb-classifier.js#L1988](https://github.com/mozilla-iot/zigbee-adapter/blob/2eaee692b067ee2d8d56531d7a6dbe9b2659d822/zb-classifier.js#L1988) only checks for RWL020.

**Comment by dhylands** [comment]
That should be an easy fix then. The input devices (i.e. remote controls) are the least discoverable of the zigbee (and zwave for that matter) devices.

**Comment by dhylands** [comment]
I've released 0.7.6 version of the zigbee adapter which added RW021 to the classifier so you should be able to get events for the dim buttons now.
